org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/model/XtextDocumentProvider.java

```diff
@@ -419,7 +428,15 @@ public boolean isSynchronized(Object element) {
 		} else {
 			return super.isSynchronized(element);
 		}
-		return synchronizationStamp == getModificationStamp(element);
+
+		// Check has to fit the following method: 
+		// org.eclipse.ui.editors.text.FileDocumentProvider.checkSynchronizationState(long, IResource)
+		boolean isSynchronized = synchronizationStamp == getSynchronizationStamp(input.getFile());
+		if (!isSynchronized) {
+			// checks for is sync and triggers sync if necessary
+			input.getFile().isSynchronized(IResource.DEPTH_ZERO);
```

Comment:
We would still return `false` though. Why is this the correct result and why is the async refresh useful here?

Reply:
IResource.isSynchornized() doesn't have the same understanding of FileDocumentProvider.checkSynchronizationState() of synchronzied.. If XtextDocumentProvider.isSynchronized is called multiple times, IResource.isSynchronized() will be for the first time unsynchron (return false) and for the next calls it will IResource.isSynchronized return true. But isSynchronized() on FileDocumentProvider will be false.. So I've to trigger it and don't care about the result.